cpp/src/arrow/compute/kernels/sort_to_indices.cc

```diff
@@ -139,6 +151,52 @@ class CountSorter {
   }
 };
 
+template <typename ArrowType, typename Comparator>
+class CountCompareSorter {
+  using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+  using c_type = typename ArrowType::c_type;
+
+ public:
+  explicit CountCompareSorter(Comparator compare) : compare_sorter_(compare) {}
+
+  void Sort(int64_t* indices_begin, int64_t* indices_end, const ArrayType& values) {
+    if (values.length() >= countsort_min_len_) {
+      auto min = std::numeric_limits<c_type>::max();
+      auto max = std::numeric_limits<c_type>::min();
+
+      for (int64_t i = 0; i < values.length(); ++i) {
+        min = values.IsNull(i) ? min : std::min(min, values.Value(i));
+        max = values.IsNull(i) ? max : std::max(max, values.Value(i));
+      }
```

Comment:
Haven't reviewed in detail yet, but this could possibly be improved by using a `BitmapReader` and a single null check

Reply:
I see `BitmapReader` for null checking in minmax kernel. Will study and update the code.

Reply:
For best performance, you actually want to use `ArrayDataVisitor`.

Reply:
@pitrou, @wesm,
I benchmarked BitmapReader and ArrayDataVisitor.
Comparing with this patch, I see big performance drop with ArrayDataVisitor. No obvious difference observed with BitmapReader. Some bechmark results are pasted below.

- ArrayDataVisitor overhead is big for counting sort
```
case: int64 within -100 ~ 100, 32K size, null = 1%
this patch: 525.675M/s
BitmapReader: 525.009M/s
ArrayDataVisitor: 401.866M/s

case: int64 within -100 ~ 100, 8M size, null = 1%
this patch: 540.892M/s
BitmapReader: 545.825M/s
ArrayDataVisitor: 412.722M/
```

- ArrayDataVisitor overhead is trivial for std::stable_sort:
```
case: int64 big range, 32K size, null = 1%
this patch: 77.6597M/s
BitmapReader: 77.9934M/s
ArrayDataVisitor: 77.5122M/s

case: int64 big range, 8M size, null = 1%
this patch: 39.9015M/s
BitmapReader: 39.7658M/
ArrayDataVisitor: 39.3367M/s
```

I guess compiler is not smart enough to inline everything in [this tight loop](https://github.com/apache/arrow/blob/dbf63796ec27c2b00aeda13d42ad423f873b3013/cpp/src/arrow/visitor_inline.h#L179-L187), which introduces too much overhead. Please review my ArrayDataVisitor test code in case I missed anything: https://pastebin.com/kjkxcZHw

My updated code is basically inlining what ArrayDataVisitor does.

Reply:
> Please review my ArrayDataVisitor test code in case I missed anything: https://pastebin.com/kjkxcZHw

Yes, it looks right. Which compiler are you using? Silly question, but you're testing in release mode, right?

Reply:
I'm using gcc7.4 on ubuntu 18.04. I tested in release mode.

P.S., I did do silly things like benchmark with ubsan enabled, and wasted a lot time finding why the huge regression :-D

Reply:
Checking Status on the hot path certainly isn't going to work for this algorithm

Reply:
Well, ideally, the compiler should notice that the Status is always ok and optimize it out...

Reply:
Looking at the code, I think [ARROW_RETURN_NOT_OK](https://github.com/apache/arrow/blob/bf00cf1e6091f48365e0cc7e982afab2ea790df1/cpp/src/arrow/status.h#L55-L59) macro should be optimized out when its input is Status::OK(). Don't know why the compiler doesn't figure out the fact.

I tried a small change, it enables compiler to optimize correctly and push ArrayDataVisitor to good performance.

- Before change:
```
#define ARROW_RETURN_NOT_OK(status)                                   \
  do {                                                                \
    ::arrow::Status __s = ::arrow::internal::GenericToStatus(status); \
    ARROW_RETURN_IF_(!__s.ok(), __s, ARROW_STRINGIFY(status));        \
  } while (false)
```

- After change (check status.ok() at the beginning):
```
#define ARROW_RETURN_NOT_OK(status)                                     \
  do {                                                                  \
    if (!(status).ok()) {                                               \
      ::arrow::Status __s = ::arrow::internal::GenericToStatus(status); \
      ARROW_RETURN_IF_(!__s.ok(), __s, ARROW_STRINGIFY(status));        \
    }                                                                   \
  } while (false)
```

Any side effect of the original two lines of code which blocks compiler optimizer?

`GenericToStatus` function is defined [here](https://github.com/apache/arrow/blob/bf00cf1e6091f48365e0cc7e982afab2ea790df1/cpp/src/arrow/status.h#L438). It gets a a copy of our Status::OK() object.

Reply:
The problem is that the `status` parameter to the macro can be an expression with side effects, and your version will execute it twice.

Reply:
Direct assigment is of good performance
```
#define ARROW_RETURN_NOT_OK2(status)                                  \
  do {                                                                \
    ::arrow::Status __s = (status);                                   \
    ARROW_RETURN_IF_(!__s.ok(), __s, ARROW_STRINGIFY(status));        \
  } while (false)
```

Calling GenericToStatus() leads to bad performance.
```
#define ARROW_RETURN_NOT_OK(status)                                   \
  do {                                                                \
    ::arrow::Status __s = ::arrow::internal::GenericToStatus(status); \
    ARROW_RETURN_IF_(!__s.ok(), __s, ARROW_STRINGIFY(status));        \
  } while (false)

inline Status GenericToStatus(const Status& st) { return st; }
```

Looks insane to me. Anyway, I think ARROW_RETURN_NOT_OK macro code is okay.